src/endpoints/sources.js

```diff
@@ -35,28 +36,37 @@ const setSource = (sourceExtractor) => (req, res, next) => {
  * @param {pathExtractor} pathExtractor - Function to extract a dataset path from the request
  * @returns {expressMiddleware}
  */
-const setDataset = (pathExtractor) => (req, res, next) => {
-  req.context.dataset = req.context.source.dataset(pathParts(pathExtractor(req)));
+const setDataset = (pathExtractor, lookForVersion=true) => (req, res, next) => {
+  console.log("middleware::setDataset")
+  req.context.dataset = req.context.source.dataset(...pathParts(pathExtractor(req), lookForVersion));
   next();
 };
 
 
 /**
  * Generate Express middleware that redirects to the canonical path for the
- * current {@link Dataset} if it is not fully resolved.
+ * current {@link Dataset} if it is not fully resolved. This includes redirecting
+ * to a more exact version, where appropriate.
  *
  * @param {pathBuilder} pathBuilder - Function to build a fully-specified path
  * @returns {expressMiddleware}
  */
 const canonicalizeDataset = (pathBuilder) => (req, res, next) => {
+  console.log("sources.js::canonicalizeDataset")
+
   const dataset = req.context.dataset;
   const resolvedDataset = dataset.resolve();
-
   if (dataset === resolvedDataset) return next();
 
+  let versionSuffix = '';
+  if (resolvedDataset.versionDescriptor && resolvedDataset.versionDescriptor!==LATEST) {
+    versionSuffix = `@${resolvedDataset.versionDescriptor}`;
+  }
+  console.log("\tCanonicalizing to ", resolvedDataset.pathParts.join("/")+versionSuffix)
+
   const canonicalPath = pathBuilder.length >= 2
-    ? pathBuilder(req, resolvedDataset.pathParts.join("/"))
-    : pathBuilder(resolvedDataset.pathParts.join("/"));
+    ? pathBuilder(req, resolvedDataset.pathParts.join("/")+versionSuffix)
+    : pathBuilder(resolvedDataset.pathParts.join("/")+versionSuffix);
```

Comment:
> This will change the URL on the user in a way I'd think we'd want to avoid.

This was a deliberate choice - if you ask for a date for which there is no dataset then it'll redirect to a date for which there is a dataset. In the case of a date which is newer than the last known version then you'll redirect to a URL without a date (i.e. the latest version). 

My thinking was that it's helpful to the user to have the URL represent the actual resource. Perhaps it's not. I'll ask more widely for feedback here.

Reply:
Nod. It still seems more confusing than useful to me, but soliciting feedback from the team as you've done is good. Even better if we could solicit feedback from users outside the core team…

Part of my reasoning here is that the date we're using in the URL and for versions is not actually a "data updated on" date. It's a "dataset file uploaded" date. There's not a strong connection to the data actually being new/updated. For example, it could actually be a reversion that makes the data _older_ than the day before.

That's why I think the right way to treat this/think about it is as "what would I have seen on this date?", and that use case doesn't suggest changing the URL to me.

If we were indexing by `meta.updated` and using those dates, that'd be a different story, and I'd think we'd want to extend (but not replace) this version access that way in the future.

Reply:
As per wider discussions this functionality has now been removed and no redirecting is performed based on the version descriptor in the URL